### DISCUSSION TITLE
Is the output of `dump()` guaranteed to be deterministic across versions?

### DISCUSSION THREAD

**Discussion by thufschmitt** [nlohmann/json#2526] [Q&A] [answered]
[The Nix package manager](https://github.com/NixOS/nix) is using this library quite a lot internally (and it's awesome btw), and we've been considering using it in a few more places. In some of these places binary reproducibility is very important − we must be sure that serializing the same json document with say `nlohmann::json` 3.9.5 and an hypothethical `nlohmann::json` 5.7.2 will yield the exact same file.

That seems to hold so far, but I couldn't find any statement guarantying this in the future. Is this actually guaranteed (except for possible implementation bugs that would have to be fixed of course)?

I'm not a json expert, but I guess that stuff like the ordering of objects or the formatting of numbers (or just whitespace) could change from one version to another without changing the semantics of the produced json.

**Comment by nlohmann** [comment, marked as answer]
The output of the `dump` function should be stable between all versions, as long as you pass the same value of indentation.

The only way where the output may change is the representation of numbers. For instance, the JSON values

```json
{"number":1000}
```

```json
{"number":1E3}
```

are equivalent. However, in recent versions, the output remained stable. But we do not give guarantees on the written bytes, but only that the output remains equivalent.

I will need to add this to the documentation somewhere.

**Reply by thufschmitt** [reply]
Thanks a lot for the quick answer.

Yes indeed, numbers were one of my fears. The order of record fields too, but since it's customizable I guess it should always be possible to get back the current behavior even if the default were to change.

Looking at the source it actually looks like the type for numbers is customizable too, so I guess that could be used to customize their pretty-printing if needs be.

**Reply by nlohmann** [reply]
To be honest - there is little you can influence on the numbers. The way a number is dumped is hard-coded in the library. Anyhow, there is no plan to change anything at this time.
